File: frontend/services/learningEngine.ts

import { ModelWeights, TradePosition, FactorBreakdown } from '../types';

/**
 * AI LEARNING ENGINE
 * 
 * This engine implements a feedback loop for the Quant Model.
 * It analyzes closed trades and adjusts the weights of the prediction factors
 * (Trend, Momentum, Delta, OrderBook) based on success or failure.
 */

export class AILearningEngine {
    private weights: ModelWeights;
    private learningRate: number = 0.05;
    private generation: number = 0;
    private history: { generation: number, result: 'WIN' | 'LOSS', weights: ModelWeights }[] = [];

    constructor() {
        // Initial Default Weights (Sum to ~1.0)
        this.weights = {
            trend: 0.40,
            momentum: 0.25,
            delta: 0.15,
            orderBook: 0.15,
            volatility: 0.05
        };
    }

    public getWeights(): ModelWeights {
        return { ...this.weights };
    }

    public getGeneration(): number {
        return this.generation;
    }

    /**
     * The Core Learning Step
     * Called when a Prediction Trade is closed.
     */
    public learn(trade: TradePosition) {
        if (trade.source !== 'PREDICTION' || !trade.metadata?.factorBreakdown) return;

        const breakdown = trade.metadata.factorBreakdown as FactorBreakdown;
        const isWin = trade.pnl > 0;
        const tradeDir = trade.side === 'LONG' ? 1 : -1;

        // "Backpropagation" Simulation
        // We look at which factors agreed with the trade direction.
        // If WIN: Boost factors that agreed.
        // If LOSS: Penalize factors that agreed (they were wrong).

        const adjustments: Partial<ModelWeights> = {};

        // Helper to check alignment
        // factorScore is the raw contribution (e.g. +50 for Uptrend).
        const checkFactor = (factorScore: number, key: keyof ModelWeights) => {
            const factorDir = factorScore > 0 ? 1 : factorScore < 0 ? -1 : 0;
            
            if (factorDir === 0) return; // Factor was neutral, no learning

            const agreed = factorDir === tradeDir;

            if (isWin) {
                // Reinforce: If it agreed, increase weight. If it disagreed, decrease weight (it was fighting the win).
                if (agreed) this.weights[key] += this.learningRate;
                else this.weights[key] -= this.learningRate * 0.5;
            } else {
                // Penalize: If it agreed (led us to loss), decrease weight. If it disagreed (was right!), increase weight.
                if (agreed) this.weights[key] -= this.learningRate;
                else this.weights[key] += this.learningRate * 0.5;
            }
        };

        checkFactor(breakdown.trend, 'trend');
        checkFactor(breakdown.momentum, 'momentum');
        checkFactor(breakdown.delta, 'delta');
        checkFactor(breakdown.orderBook, 'orderBook');
        // Volatility is usually a scaler, not directional, so we might skip or treat differently. 
        // For simplicity, let's treat it as a confidence booster.

        this.normalizeWeights();
        this.generation++;
        
        this.history.push({
            generation: this.generation,
            result: isWin ? 'WIN' : 'LOSS',
            weights: { ...this.weights }
        });

        console.log(`[AI LEARNING] Gen ${this.generation} | Result: ${isWin ? 'WIN' : 'LOSS'} | Updated Weights:`, this.weights);
    }

    private normalizeWeights() {
        const total = Object.values(this.weights).reduce((a, b) => a + Math.abs(b), 0);
        if (total === 0) return;

        // Normalize to sum to 1.0
        (Object.keys(this.weights) as Array<keyof ModelWeights>).forEach(key => {
            this.weights[key] = Math.max(0.01, this.weights[key] / total); // Prevent 0 weight
        });
    }
}

File: backend_backup_20260319_124426/app/domain/fabio_ai/services/learning_engine.py
"""Learning Engine — adaptive weight tuning domain service.

Implements a simplified 'backpropagation' loop that adjusts prediction model
weights based on trade outcomes, reinforcing factors that align with wins
and penalising those that align with losses.
"""

from __future__ import annotations

from app.domain.trading.models.enums import Side, Source
from app.domain.trading.models.entities import Position
from app.domain.fabio_ai.models.predictions import ModelWeights, FactorBreakdown


class LearningEngine:
    """Adaptive weight tuning engine.

    Adjusts weights of the PredictionEngine's scoring factors based on
    closed PREDICTION trades.
    """

    def __init__(self, learning_rate: float = 0.05) -> None:
        self._weights = {
            "trend": 0.40,
            "momentum": 0.25,
            "delta": 0.15,
            "order_book": 0.15,
            "volatility": 0.05,
        }
        self._learning_rate = learning_rate
        self._generation = 0
        self._history: list[dict] = []

    @property
    def weights(self) -> ModelWeights:
        return ModelWeights(**self._weights)

    @property
    def generation(self) -> int:
        return self._generation

    def learn(self, trade: Position) -> None:
        """Process a closed PREDICTION trade and adjust weights."""
        if trade.source != Source.PREDICTION:
            return
        if not trade.metadata or "factorBreakdown" not in trade.metadata:
            return

        raw = trade.metadata["factorBreakdown"]
        # raw may be a dict (from JSON) or already a FactorBreakdown
        if isinstance(raw, FactorBreakdown):
            bd = raw
        elif isinstance(raw, dict):
            bd = FactorBreakdown(**raw)
        else:
            return

        is_win = trade.pnl > 0
        trade_dir = 1 if trade.side == Side.LONG else -1

        self._adjust_factor(bd.trend, "trend", is_win, trade_dir)
        self._adjust_factor(bd.momentum, "momentum", is_win, trade_dir)
        self._adjust_factor(bd.delta, "delta", is_win, trade_dir)
        self._adjust_factor(bd.order_book, "order_book", is_win, trade_dir)

        self._normalize()
        self._generation += 1

        self._history.append({
            "generation": self._generation,
            "result": "WIN" if is_win else "LOSS",
            "weights": dict(self._weights),
        })

    def _adjust_factor(
        self, factor_score: float, key: str, is_win: bool, trade_dir: int
    ) -> None:
        if factor_score == 0:
            return

        factor_dir = 1 if factor_score > 0 else -1
        agreed = factor_dir == trade_dir

        if is_win:
            if agreed:
                self._weights[key] += self._learning_rate
            else:
                self._weights[key] -= self._learning_rate * 0.5
        else:
            if agreed:
                self._weights[key] -= self._learning_rate
            else:
                self._weights[key] += self._learning_rate * 0.5

    def _normalize(self) -> None:
        total = sum(abs(v) for v in self._weights.values())
        if total == 0:
            return
        for key in self._weights:
            self._weights[key] = max(0.01, self._weights[key] / total)
